core/embed/trezorhal/stm32f4/flash.c

```diff
@@ -104,48 +110,165 @@ const void *flash_get_address(uint8_t sector, uint32_t offset, uint32_t size) {
   return (const void *)addr;
 }
 
-uint32_t flash_sector_size(uint8_t sector) {
+uint32_t flash_sector_size(uint16_t sector) {
   if (sector >= FLASH_SECTOR_COUNT) {
     return 0;
   }
   return FLASH_SECTOR_TABLE[sector + 1] - FLASH_SECTOR_TABLE[sector];
 }
 
-secbool flash_erase_sectors(const uint8_t *sectors, int len,
-                            void (*progress)(int pos, int len)) {
+static uint32_t flash_subarea_get_size(const flash_subarea_t *subarea) {
+  uint32_t size = 0;
+  for (int s = 0; s < subarea->num_sectors; s++) {
+    size += flash_sector_size(subarea->first_sector + s);
+  }
+  return size;
+}
+
+uint32_t flash_area_get_size(const flash_area_t *area) {
+  uint32_t size = 0;
+  for (int i = 0; i < area->num_subareas; i++) {
+    size += flash_subarea_get_size(&area->subarea[i]);
+  }
+  return size;
+}
+
+uint16_t flash_total_sectors(const flash_area_t *area) {
+  uint16_t total = 0;
+  for (int i = 0; i < area->num_subareas; i++) {
+    total += area->subarea[i].num_sectors;
+  }
+  return total;
+}
+
+int32_t flash_get_sector_num(const flash_area_t *area,
+                             uint32_t sector_inner_num) {
+  uint16_t sector = 0;
+  uint16_t remaining = sector_inner_num;
+  for (int i = 0; i < area->num_subareas; i++) {
+    if (remaining < area->subarea[i].num_sectors) {
+      sector = area->subarea[i].first_sector + remaining;
+      return sector;
+    } else {
+      remaining -= area->subarea[i].num_sectors;
+    }
+  }
+
+  return -1;
+}
+
+static secbool subarea_get_sector_and_offset(const flash_subarea_t *subarea,
+                                             uint32_t offset,
+                                             uint16_t *sector_out,
+                                             uint32_t *offset_out) {
+  uint32_t tmp_offset = offset;
+  uint16_t sector = subarea->first_sector;
+
+  // in correct subarea
+  for (int s = 0; s < subarea->num_sectors; s++) {
+    const uint32_t sector_size = flash_sector_size(sector);
+    if (tmp_offset >= sector_size) {
+      tmp_offset -= sector_size;
+      sector++;
+
+      if (s == subarea->num_sectors - 1) {
+        // last sector in subarea
+        return secfalse;
+      }
+      continue;
+    }
+    *sector_out = sector;
+    *offset_out = tmp_offset;
+    return sectrue;
```

Comment:
This code seems more complicated than it needs to be. I think we can make it easier to read, or am I overlooking something?
```suggestion
    if (tmp_offset < sector_size) {
      *sector_out = sector;
      *offset_out = tmp_offset;
      return sectrue;
    }
    tmp_offset -= sector_size;
    sector++;
```

Reply:
yeah this is better https://github.com/trezor/trezor-firmware/commit/e9e84cc14873a636045edb6d4c2a6623f333d355